cli/loader_lib.c

```diff
@@ -153,6 +153,177 @@ JL_DLLEXPORT const char * jl_get_libdir()
     return lib_dir;
 }
 
+#ifdef _OS_LINUX_
+#define JL_DEFAULT_STDCXXLIB "libstdc++.so.6"
+#ifndef GLIBCXX_LEAST_VERSION /* Should always be defined in the makefile. This appeases the linter. */
+#define GLIBCXX_LEAST_VERSION "GLIBCXX_3.4.29"
+#endif
+
+#include <link.h>
+#include <sys/wait.h>
+
+// On Linux, it can happen that the system has a newer libstdc++ than the one we ship,
+// which can break loading of some system libraries: <https://github.com/JuliaLang/julia/issues/34276>.
+
+// Return the path to the libstdcxx to load. Otherwise, print the error and exit.
+// The path returned must be freed.
+static char* libstdcxxprobe(void)
+{
+    // Create the pipe and child process.
+    int fork_pipe[2];
+    int ret = pipe(fork_pipe);
+    if (ret == -1) {
+        fputs("Error during libstdcxxprobe:", stderr);
+        perror("pipe");
+        fflush(stderr);
+        exit(1);
+    }
+    // Make sure the buffers are clean before forking.
+    fflush(stdout);
+    fflush(stderr);
+    pid_t pid = fork();
+    if (pid == -1)  {
+        fputs("Error during libstdcxxprobe:", stderr);
+        perror("fork");
+        fflush(stderr);
+        exit(1);
+    }
+    if (pid == (pid_t) 0) { // Child process.
+        ret = close(fork_pipe[0]);
+        if (ret == -1) {
+            fputs("Error during libstdcxxprobe in child process:", stderr);
+            perror("close");
+            fflush(stderr);
```

Comment:
Let's avoid getting our output mixed up, reversed, backwards, and duplicated with the parent process by avoiding libc's `FILE *stderr` handle.
```suggestion
            perror("Error during libstdcxxprobe in child process:\nclose");
```